### PR TITLE
PP 7 registration

### PR DESCRIPTION
**Pull request by ofeterovskaya**
Student can register
Student_Registration
if user is already exists, he got a notification "User already exists"
Register_theSameUser
if user is <16yo, user needs to provide adult name
studentLess16Registered
if user doesn't provide adult name, he/she got an error
UserIsLess15
Teacher is able to register 
teacherRegister
if the same teacher registers again, he got an error "user already exists"
teacherExists

### CODE REVIEW

**Review by benochi — CHANGES_REQUESTED** [review]
Consolidate Registration into One Controller:

Combine registerUser, registerTeacher, and registerStudent into a single controller to handle all types of user registrations.
You can have a general Auth controller and conditional blocks so we don't have repetitive controllers. 


Avoid Redundant Hashing:
Ensure password hashing occurs only in the pre('save') hook in the User model, not in the controller.

Hashing Only on New Creation or Password Change:
Modify the pre('save') hook in the User model to hash the password only if it has been modified.

IE: 
UsersSchema.pre('save', async function(next) {
  if (!this.isModified('password')) return next();
  
  const salt = await bcrypt.genSalt(10);
  this.password = await bcrypt.hash(this.password, salt);
  next();
});

File: src/app.js [hunk 1]
```diff
@@ -24,6 +24,7 @@ app.use(express.static('public'))
 app.use(favicon(__dirname + '/public/favicon.ico'));
 
 // routes
-app.use('/api/v1', mainRouter, userRouter, classesRouter);
+app.use('/api/v1', mainRouter, userRouter , classesRouter);
```
Comment (AnnaPestova1) [inline src/app.js:27]: do we need to use here `userRouter , classesRouter` because they use `/api/v1/users` path on the line 28

Comment (ofeterovskaya) [reply src/app.js]: changed, and added letter validation, so User should enter only letters for last name and first name, and also if adultName is required, I added validation to check if it contains first AND last name

File: src/app.js [hunk 2]
```diff
@@ -24,6 +24,7 @@ app.use(express.static('public'))
 app.use(favicon(__dirname + '/public/favicon.ico'));
 
 // routes
-app.use('/api/v1', mainRouter, userRouter, classesRouter);
+app.use('/api/v1', mainRouter, classesRouter);
```
Comment (AnnaPestova1) [inline src/app.js:27]: classRouter is here and on the next line))) Where it should 'live"?

Comment (ofeterovskaya) [reply src/app.js:27]: done . in users should be only userRoutes

File: src/models/User.js [hunk 3]
```diff
@@ -1,30 +1,82 @@
-const mongoose = require("mongoose");
+const mongoose = require('mongoose');
+const bcrypt = require('bcryptjs');
+const { adultValidator, adultNameFirstAndLast } = require('../utils/adultValidation');
+const {lettersOnlyValidator}= require("../utils/letterValidation.js");
 
-const UserSchema = new mongoose.Schema({
+// Define the Users schema
+const UsersSchema = new mongoose.Schema({
```
Comment (benochi) [inline src/models/User.js:7]: It should be UserSchema since the model is User, and not Users.  Not a big deal.

File: src/models/User.js [hunk 4]
```diff
@@ -1,30 +1,82 @@
...
-const UserModel = mongoose.model("User", UserSchema, "Users");
+// Compare the provided password with the users' password
+UsersSchema.methods.comparePassword = async function(candidatePassword) {
+  const isMatch = await bcrypt.compare(candidatePassword, this.password);
+  return isMatch;
+};
 
-module.exports = UserModel;
\ No newline at end of file
+module.exports = mongoose.model('Users', UsersSchema);
```
Comment (benochi) [inline src/models/User.js:82]: And export it as  User and update any controllers using Users

**Review by benochi — CHANGES_REQUESTED** [review]
Minor typos only. 
Are we using the express-validation yet?

**Comment by ofeterovskaya** [comment]
> Minor typos only. Are we using the express-validation yet?

Yes, In our userValidation.js file, I utilize express-validator for validating user data. I use body and validationResult from express-validator to define and check the validation rules for both students and teachers. This helps ensure that the data submitted by users is properly validated before any further processing.

**Review by benochi — APPROVED** [review]

**Review by benochi — APPROVED** [review]

**Pull request merged**